File: examples_php/example_random.php
#!/usr/bin/env php
<?php

declare(strict_types=1);
require_once __DIR__ . '/Zupt.php';

try {
    $zupt = new ZuptFFI();

    echo str_repeat('=', 60), PHP_EOL;
    echo "libzupt - Aleatoriedade e hashes PHP/FFI", PHP_EOL;
    echo str_repeat('=', 60), PHP_EOL, PHP_EOL;

    echo "1. Gerando 32 bytes aleatórios...", PHP_EOL;
    $random = $zupt->randomBytes(32);
    echo '   ', bin2hex($random), PHP_EOL, PHP_EOL;

    echo "2. Gerando nonce AES...", PHP_EOL;
    $nonce = $zupt->randomBytes(ZuptFFI::AES_NONCE_SIZE);
    echo '   ', bin2hex($nonce), PHP_EOL, PHP_EOL;

    $data = 'Hello, Post-Quantum World!';
    echo "3. SHA-256 de '{$data}'...", PHP_EOL;
    echo '   ', bin2hex($zupt->sha256($data)), PHP_EOL, PHP_EOL;

    echo "4. SHA3-512 de '{$data}'...", PHP_EOL;
    echo '   ', bin2hex($zupt->sha3_512($data)), PHP_EOL, PHP_EOL;

    echo "5. Simulando derivação simples...", PHP_EOL;
    $salt = $zupt->randomBytes(16);
    $derived = $zupt->sha256($salt . 'my-secret-password');
    echo '   Salt: ', bin2hex($salt), PHP_EOL;
    echo '   Chave derivada: ', bin2hex($derived), PHP_EOL;

    echo PHP_EOL, str_repeat('=', 60), PHP_EOL;
    echo "Exemplo de aleatoriedade concluído!", PHP_EOL;
    echo str_repeat('=', 60), PHP_EOL;
} catch (Throwable $error) {
    fwrite(STDERR, 'ERRO: ' . $error->getMessage() . PHP_EOL);
    exit(1);
}

File: examples_dotnet/ExampleRandom.cs
using LibZupt;
using System.Text;

namespace ZuptExample;

internal static class ExampleRandom
{
    private static string Hex(byte[] data) => Convert.ToHexString(data).ToLowerInvariant();

    public static void Run()
    {
        Console.WriteLine(new string('=', 60));
        Console.WriteLine("libzupt - Random Bytes and Hashing Example");
        Console.WriteLine(new string('=', 60));
        Console.WriteLine();

        Console.WriteLine("1. Generating random bytes...");
        byte[] randomBytes = Zupt.RandomBytes(32);
        Console.WriteLine($"   Generated {randomBytes.Length} random bytes:");
        Console.WriteLine($"   {Hex(randomBytes)}");
        Console.WriteLine();

        Console.WriteLine("2. Generating AES nonce...");
        byte[] nonce = Zupt.RandomBytes(ZuptConstants.AesNonceSize);
        Console.WriteLine($"   Nonce ({nonce.Length} bytes): {Hex(nonce)}");
        Console.WriteLine();

        Console.WriteLine("3. Computing SHA-256 hash...");
        byte[] data = Encoding.UTF8.GetBytes("Hello, Post-Quantum World!");
        byte[] sha256Hash = Zupt.Sha256(data);
        Console.WriteLine($"   Data: {Encoding.UTF8.GetString(data)}");
        Console.WriteLine($"   SHA-256: {Hex(sha256Hash)}");
        Console.WriteLine();

        Console.WriteLine("4. Computing SHA3-512 hash...");
        byte[] sha3_512Hash = Zupt.Sha3_512(data);
        Console.WriteLine($"   Data: {Encoding.UTF8.GetString(data)}");
        Console.WriteLine($"   SHA3-512: {Hex(sha3_512Hash)}");
        Console.WriteLine();

        Console.WriteLine("5. Simulating key derivation...");
        byte[] salt = Zupt.RandomBytes(16);
        Console.WriteLine($"   Salt: {Hex(salt)}");
        byte[] derivedKey = Zupt.Sha256(Concat(salt, Encoding.UTF8.GetBytes("my-secret-password")));
        Console.WriteLine($"   Derived key (32 bytes): {Hex(derivedKey)}");
        Console.WriteLine();

        Console.WriteLine(new string('=', 60));
        Console.WriteLine("Random bytes and hashing example passed!");
        Console.WriteLine(new string('=', 60));
        Console.WriteLine();
    }

    private static byte[] Concat(byte[] a, byte[] b)
    {
        var result = new byte[a.Length + b.Length];
        a.CopyTo(result, 0);
        b.CopyTo(result, a.Length);
        return result;
    }
}